### ISSUE TITLE
Whats the baud rate and pinout on the Wio terminal?

### ISSUE THREAD

**Issue by twitchyliquid64** [Seeed-Studio/Seeed_Arduino_rpcUnified#6]
Hi!

I'm implementing support for Wio Terminal + WiFi (using eRPC ameba firmware) in Rust.

Can you help me work out what the parameters are for talking to the ameba? In particular, I'm not sure what the baud rate is.

Is this correct?
 - Reset the RTL8720 by bringing `PA18` low for 100ms, then high, and waiting 100ms
 - Communicate using NXP eRPC over  `SERCOM0` in UART mode, with RX on `PC24` (Pad2) and TX on `PB24` (Pad 0). No CTS/RTS.
 - ? Baud rate 1843200 ?
 - eRPC uses the default codec, default framed transport, default crc16 (starting value `0xEF4A`, polynomial `0x1021`)

**Comment by lyzsuper** [comment]
Hi,Twitch

1.yes,use PA18 at WIO Terminal to reset RTL8720,100ms low and 100ms high
2.yes,use  SERCOM0,PC24(PAD2)/PB24(PAD0),no CTS/RTS at WIO Terminal
3.yes,use badu rate at 1843200
4.yes,default codec, default framed transport, default crc16(crcStart is 0xEF4A,polynomial is 0x1021)

**Comment by twitchyliquid64** [comment]
Got my eRPC implementation working! :DD

One thing to note: I'm observing the Ameba is taking at least 120ms to be ready after I bring `PA18` high, so any bits sent earlier are corrupted.

Maybe increase the delay before any bits are written to it? I'm having no issues if I give it 200ms.

**Comment by Pillar1989** [comment]
@twitchyliquid64  We are very glad that our software can help you. Can you share your rust repository? I'm very interested.

**Comment by twitchyliquid64** [comment]
@Pillar1989 I'm implementing this upstream in https://github.com/atsamd-rs/atsamd, specifically in the Wio Terminal board support crate ([here](https://github.com/atsamd-rs/atsamd/tree/master/boards/wio_terminal)).

My eRPC / Wifi implementation is a work in progress on my local machine, however I hope to get it merged upstream into atsamd-rs in the next week or so.

**Comment by Pillar1989** [comment]
@twitchyliquid64  Great. I'm really looking forward to see eRPC run successfully on the rust.

**Comment by twitchyliquid64** [comment]
Still a work in progress but ive gotten the `GetSystemVersion` & `GetMacAddress` RPCs working.

Main controller: https://github.com/twitchyliquid64/atsamd/blob/wio_terminal/boards/wio_terminal/src/wifi.rs#L54
RPC trait: https://github.com/twitchyliquid64/atsamd/blob/wio_terminal/boards/wio_terminal/src/wifi/erpc.rs#L165
RPC implementation for `GetMacAddress`: https://github.com/twitchyliquid64/atsamd/blob/wio_terminal/boards/wio_terminal/src/wifi/wifi_rpcs.rs#L5

Aside from implementing the bulk of the RPCs, im not 100% happy with the RPC trait or the streaming parser yet, so it will probably be a week or two before its upstream and ready for public use.

**Comment by Pillar1989** [comment]
@twitchyliquid64    Is it possible to write a rust erpcgen(https://github.com/EmbeddedRPC/erpc/tree/develop/erpcgen)? This will automatically generate the corresponding code from an idl file.
  https://github.com/Seeed-Studio/seeed-ambd-firmware/tree/master/erpc_idl
Most of our code here is generated automatically by erpcgen.
https://github.com/Seeed-Studio/Seeed_Arduino_rpcUnified/tree/master/src

**Comment by twitchyliquid64** [comment]
Yes, there are some corner cases (because I am trying to do everything without memory allocations) however.

**Comment by twitchyliquid64** [comment]
I've published a crate for this: for now it only supports scanning RPCs + getting the MAC + getting the version: https://github.com/twitchyliquid64/seeed-erpc-rs

This crate only handles parsing/serializing the different RPCs, so it should be usable from any Rust embedded codebase. It also does everything entirely on the stack, so no memory allocator is needed.

Next up I'll implement connecting to APs, and some of the LWIP RPCs. Should be straightforward now I have all the boilerplate working.

I'm also merging an [example](https://github.com/atsamd-rs/atsamd/pull/353/files#diff-d1966ad53cf2a08cc6f08dcf26bd5287276339e5fe990be2945f111593fb1230) that shows all the wifi networks on the screen, into the atsamd HAL.

**Comment by Pillar1989** [comment]
@twitchyliquid64  rust eRPC have any new progress?

**Comment by twitchyliquid64** [comment]
@Pillar1989 at the moment just AP scanning and connecting to a network is implemented, but next I want to get UDP sockets working. I'll hopefully have that in sometime this week.
